addons/bus/static/src/js/crosstabbus.js

```diff
@@ -0,0 +1,350 @@
+        if ((parseInt(heartbeatValue) + this.heartbeatOutOfDatePeriod) < current) {
+            // Heartbeat is out of date. Electing new master
+            this._startElection();
+            heartbeatValue = this._callLocalStorage('getItem', 'heartbeat') || 0;
+        }
+
+        if (this._isMasterTab) {
+            //walk through all peers and kill old
+            var cleanedPeers = {};
+            for (var peerName in peers) {
+                if (peers[peerName] + this.heartbeatKillOldPeriod > current) {
+                    cleanedPeers[peerName] = peers[peerName];
+                }
+            }
+
+            if (parseInt(heartbeatValue) !== this.lastHeartbeat) {
+                // someone else is also master...
+                // it should not happen, except in some race condition situation.
+                this._isMasterTab = false;
+                this.lastHeartbeat = 0;
+                peers[this.id] = current;
+                this._callLocalStorage('setItem', 'peers', peers);
+                this.stopPolling();
+                this.trigger('no_longer_master');
+            } else {
+                this.lastHeartbeat = current;
+                this._callLocalStorage('setItem', 'heartbeat', current);
+                this._callLocalStorage('setItem', 'peers', cleanedPeers);
+            }
+        } else {
+            //update own heartbeat
+            peers[this.id] = current;
+            this._callLocalStorage('setItem', 'peers', peers);
+        }
+
+        // Write lastPresence in local storage if it has been updated since last heartbeat
+        var hbPeriod = this._isMasterTab ? this.masterTabHeartbeatPeriod : this.tabHeartbeatPeriod;
+        if (this._lastPresence + hbPeriod > new Date().getTime()) {
+            this._callLocalStorage('setItem', 'lastPresence', this._lastPresence);
+        }
+
+        this._heartbeatTimeout = setTimeout(this._heartbeat.bind(this), hbPeriod);
+    },
+    /**
+     * Check with the local storage if the current tab is the master tab.
+     * If this tab became the master, trigger 'become_master' event
+     *
+     */
+    _startElection: function () {
+        if (this._isMasterTab) {
+            return;
+        }
+        //check who's next
+        var peers = this._callLocalStorage('getItem', 'peers', {});
+        var now = new Date().getTime();
+        var newMaster;
+        for (var peerName in peers) {
+            //check for dead peers
+            if (peers[peerName] + this.heartbeatKillOldPeriod < now) {
+                continue;
+            }
+
+            newMaster = peerName;
+            break;
+        }
+
+        if (newMaster === this.id) {
+            //we're next in queue. Electing as master
+            this.lastHeartbeat = new Date().getTime();
+            this._callLocalStorage('setItem', 'heartbeat', this.lastHeartbeat);
+            this._isMasterTab = true;
+            this.startPolling();
+            this.trigger('become_master');
```

Comment:
What is the purpose of this trigger? I do not see any `.on('become_master', ...)`